File: python/2020/02_password_philosophy/aoc202002.py
"""AoC 2, 2020: Password Philosophy."""

# Standard library imports
import pathlib
import sys


def parse_data(puzzle_input):
    """Parse input."""
    return [parse_policy(line) for line in puzzle_input.split("\n")]


def parse_policy(line):
    """Parse one line of input.

    ## Example:

    >>> parse_policy("3-10 o: adventofcode")
    (3, 10, 'o', 'adventofcode')
    """
    numbers, char, password = line.split()
    first, second = numbers.split("-")
    return (int(first), int(second), char[0], password)


def part1(policies):
    """Solve part 1."""
    return sum(is_valid_count(*policy) for policy in policies)


def part2(policies):
    """Solve part 2."""
    return sum(is_valid_position(*policy) for policy in policies)


def is_valid_count(first, second, char, password):
    """Check if the password follows the count requirements.

    ## Examples:

    >>> is_valid_count(4, 6, "e", "adventofcode")
    False

    >>> is_valid_count(1, 3, "o", "passwordphilosophy")
    True
    """
    return first <= password.count(char) <= second


def is_valid_position(first, second, char, password):
    """Check if the password follows the position requirements.

    ## Examples:

    >>> is_valid_position(4, 6, "e", "adventofcode")
    True

    >>> is_valid_position(1, 3, "o", "passwordphilosophy")
    False
    """
    return (password[first - 1] == char) != (password[second - 1] == char)


def solve(puzzle_input):
    """Solve the puzzle for the given input."""
    data = parse_data(puzzle_input)
    yield part1(data)
    yield part2(data)


if __name__ == "__main__":
    for path in sys.argv[1:]:
        print(f"\n{path}:")
        solutions = solve(puzzle_input=pathlib.Path(path).read_text().strip())
        print("\n".join(str(solution) for solution in solutions))

File: julia/2020/02_password_philosophy/aoc202002.jl
# Advent of Code 2020, day 2: Password Philosophy

module AOC202002

# Parse each line into parts: "{n1}-{n2} {char}: {word}"
function parse_line(line)
    pattern = r"^(?P<n1>\d+)-(?P<n2>\d+) (?P<char>.): (?P<word>.+)$"
    parts = match(pattern, line)
    Dict(
        "num_1" => parse(Int16, parts[:n1]),
        "num_2" => parse(Int16, parts[:n2]),
        "char" => parts[:char][1],
        "word" => parts[:word],
    )
end


# Check password based on first policy: the given character must occur between
# num_1 and num_2 times.
function is_valid_count(parts)
    char, word = parts["char"], parts["word"]
    parts["num_1"] <= count(c -> (c == char), word) <= parts["num_2"]
end


# Check password based on second policy: Exactly one of characters at indices
# num_1 and num_2 must match the given character.
function is_valid_position(parts)
    char, word = parts["char"], parts["word"]
    (word[parts["num_1"]] == char) ⊻ (word[parts["num_2"]] == char)
end

# Parse input
function parse_data(puzzle_input)
    split(puzzle_input, "\n") .|> parse_line
end


# Solve part 1
function part1(data)
    data .|> is_valid_count |> sum
end

# Solve part 2
function part2(data)
    data .|> is_valid_position |> sum
end

# Solve the puzzle for the given input
function solve(puzzle_input)
    data = puzzle_input |> parse_data
    part1(data), part2(data)
end


# Solve the problem for one file
function solve_file(file_path)
    "\n$(file_path)" |> println
    open(file_path) do fid
        read(fid, String) |> strip .|> solve
    end
end

# Solve the problem for each file
[a for a in ARGS if a[1] != '-'] .|> solve_file .|> s -> join(s, "\n") |> println

end  # module
